### DISCUSSION TITLE
Docs recommend OAuth2 password flow / RFCS 6749 is against it

### DISCUSSION THREAD

**Discussion by msander** [fastapi/fastapi#9263] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# no code, just a comment about the docs
```

### Description

The OAuth2 RFC contains the following warning about the "Resource Owner Password Credentials Grant"

>The authorization server should take special care when enabling this grant type and only allow it when other flows are not viable.

FastAPI docs are very beginner-friendly and "recommend" the password flow as a suitable method. From security perspective, this flow creates huge risks for a real Identity Provider: One compromised application can compromise all user passwords and thus all applications. This does not happen with the authorization code flow.
I feel the docs are leading beginners in a wrong direction. 

### Operating System

Linux, macOS

### FastAPI Version

*

### Python Version

*

**Comment by ErikvdVen** [comment]
I believe it really depends on the kind of application you are building. For me, for example, I'm using FastAPI but neither build a SPA application nor allow other third party (web) applications to access my back-end. I render pages statically and use Jinja as a template engine. In this case the password flow kind of makes sense to use.

However I do agree that it would be nice to inform users who actually do allow other apps to access the API or use a SPA-like application, to inform them about the Authorization flow which, if I recall correctly, redirects users to a static rendered secured webpage (on a subdomain for example) which returns the user back to the SPA application after authentication with an Authorization key to use. But that probably should be an advanced topic.

**Reply by msander** [reply]
Why are you using OpenID Connect with a single application? I think the typical use case is either an enterprise single-sign-on environment with xx applications or social login. 

Microsoft warns about this:

>Microsoft recommends you do not use the ROPC flow. In most scenarios, more secure alternatives are available and recommended. This flow requires a very high degree of trust in the application, and carries risks that are not present in other flows. **You should only use this flow when other more secure flows aren't viable.**

Okta warns about this:

> The Resource Owner Password flow is not a recommended approach. It is intended for applications for which no other flow works, as it requires your application code to be fully trusted and protected from credential-stealing attacks. It is made available primarily to provide a consistent and predictable integration pattern for legacy applications that can't otherwise be updated to a more secure flow such as the Authorization Code flow. **This should be your last option, not your first choice.**

Auth0 warns about this:

> Though we do not recommend it, highly-trusted applications can use the Resource Owner Password Flow (defined in [OAuth 2.0 RFC 6749, section 4.3](https://tools.ietf.org/html/rfc6749#section-4.3)), which requests that users provide credentials (username and password), typically using an interactive form. Because credentials are sent to the backend and can be stored for future use before being exchanged for an Access Token, it is imperative that the application is absolutely trusted with this information.
> 
> Even if this condition is met, the Resource Owner Password Flow **should only be used when redirect-based flows (like the [Authorization Code Flow](https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow)) cannot be used**.

Keycloak warns about this:

> It’s not recommended to use this flow **unless you absolutely need to**. Examples where this could be useful are legacy applications and command-line interfaces.

**Comment by ErikvdVen** [comment]
I'm not saying I am using an authentication method like that. My own application will be only reachable by VPN and I still have to implement an authentication method, which is mainly to differentiate between the users. But that's another topic. 

Ok, so it is more secure to use something as keycloak instead of ROPC. And it is like you said: if the more secure options are available to you. I believe, in that case, it would just be better for you to open a request: https://github.com/tiangolo/fastapi/issues/new/choose or a PR with your proposal.

**Reply by msander** [reply]
Seems like only @tiangolo is allowed to create issues :-)

**Comment by msander** [comment]
@tiangolo If you agree on this topic and let me create ticket, I will create a pull request which at least includes a warning.

**Comment by ZeroAurora** [comment]
Upvoting this. Using OAuth in this case is more misleading than helpful, since the so-called "password grant" doesn't cover authentication according to OAuth 2.0 spec. Furthermore, this flow isn't valid any more in OAuth 2.1.
I think that password grant may be used as an demonstration but warnings must be there to urge new developers to learn authentication technologies like OpenID Connect.
BTW some personal ideas: yet in real world we do see usernames and passwords flowing over APIs to login and get access code. That sounds more like an internal thing, an API-Key exchange or something, not an "OAuth" thing or whatever.

**Comment by justinbrick** [comment]
Agree wholeheartedly, and then some.
The entire piece of documentation reeks of bad practice. It needs an entire rewrite. It's **usually** not the goal of the API author to act as an authorization server and should rather include documentation on how to validate the integrity of an access token through an introspection endpoint or with JWT validation.

I feel the current documentation is dubious and harmful to new developers that might not know any better. What it should contain is the best practices for API validation, and in the off-chance that the API **really** needs to store the access tokens, a guide for using the PKCE flow, and **not** the implicit flow, as this is also not recommended anymore.

I wish to contribute on this topic, but I'm not a web developer, and don't understand the structure of the docs folder. Is this something that requires modifications from @tiangolo? 
If this is a scenario where we are able to open a PR to change the documentation ourselves, I would be more than happy to give an outline in a document so that someone more experienced than me on docs maintenance can put it onto the docs page.
